### DISCUSSION TITLE
Start off with curl_multi_socket_action instead of curl_multi_socket_all

### DISCUSSION THREAD

**Discussion by ThePedroo** [curl/curl#13180] [Q&A]
Hello,

I've been working on replacing `curl_multi_socket_all` with `curl_multi_socket_action`, as it is deprecated. However, when replacing it, the connection seems to simply not proceed.

Before that, a `curl_multi_perform` was introduced after to match [curl_multi_socket_all behavior inside multi_socket function](https://github.com/curl/curl/blob/master/lib/multi.c#L3131), however this introduced high CPU usage.

The [source code is available in GitHub](https://github.com/Cogmasters/concord/blob/1f2a9581e0c0ca31549de87296d53bb4ba7eae78/core/websockets.c#L828-L848) if necessary.

I appreciate all work done to make cURL.

OBS: That's not the only part that does that when replacing for _socket, however a solution to one of them should solve all of them.

**Comment by bagder** [comment]
Can you describe the situation when you are calling this function?

**Reply by ThePedroo** [reply]
There are 3 cases where they're used, where one of them I can't really understand due to the complexity of the library, but I believe all of them should fall into the same case. Those are used to initialize connections, where if you remove them, they won't proceed with the processing of it (mostly read).

**Comment by ThePedroo** [comment]
Can this have some attention? It's rather important to have more information on how to switch from `*_socket_action`, as specifically for it, there isn't much information about.

**Reply by bagder** [reply]
I don't understand the question.

After you've added one or more easy handles, you can kick everything off by calling curl_multi_socket_action() with `CURL_SOCKET_TIMEOUT` set.

**Reply by ThePedroo** [reply]
You are correct, it does work. However it utilizes an unreasonable amount of CPU.

For reference, before moving to `curl_multi_socket_action`, it hardly ever had utilized more than 1% of my machine's CPU, which is an Intel i5 2rd gen, after the changes, it easily uses more than 40%. Those were not only reproducible by me, but other users utilizing the development branch, which had the changes.

The changes are available in GitHub if needed: https://github.com/Cogmasters/concord/commit/1f2a9581e0c0ca31549de87296d53bb4ba7eae78

**Reply by bagder** [reply]
That change looks totally insane to me in the way it mixes `curl_multi_socket_action` and `curl_multi_perform()`. I cannot fathom how that is supposed to actually work and I have never seen anyone do it like that before.

It rather looks to me like someone is not quite understanding how this API works.

**Reply by ThePedroo** [reply]
This is not about "not quite understanding how this API works". There isn't any guide/documentation explaining how to switch to `curl_multi_socket_action`, making this rather harder to work on.

There's still the old code using `curl_multi_socket_all` which can be seen, but it's not much different from the "current" code, so I believe it won't help you much to understand what it is doing.

> That change looks totally insane to me in the way it mixes `curl_multi_socket_action` and `curl_multi_perform()`. 

The addition of `curl_multi_perform` is due to the issue where the sockets wouldn't get initialized/run without it.

It would useful if you informed how to replace `curl_multi_socket_all` with `curl_multi_socket_action`, as although there is for `curl_multi_socket`, there isn't for `curl_multi_socket_all` -- and if it can replaced by non-all function, it would be even better.

**Reply by bagder** [reply]
I think we might make better progress here if you asked about what specific problems you have or what details about `curl_multi_socket_action` you don't understand. Switching from multi_perform() to multi_socket is a pretty big switch, mostly because switching anything from a simple main loop into an event-based system changes everything because of the different paradigm. 

> It would useful if you informed how to replace curl_multi_socket_all with curl_multi_socket_action

I honestly can't because I don't understand when or why you need `curl_multi_socket_all` to begin with. With an event-based system, there should be no need to drive all transfers at once unconditionally.
